aws/resource_aws_appsync_api_key.go

```diff
@@ -0,0 +1,164 @@
+package aws
+
+import (
+	"fmt"
+	"log"
+	"strings"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/appsync"
+	"github.com/hashicorp/terraform/helper/schema"
+	"github.com/hashicorp/terraform/helper/validation"
+)
+
+func resourceAwsAppsyncApiKey() *schema.Resource {
+
+	return &schema.Resource{
+		Create: resourceAwsAppsyncApiKeyCreate,
+		Read:   resourceAwsAppsyncApiKeyRead,
+		Update: resourceAwsAppsyncApiKeyUpdate,
+		Delete: resourceAwsAppsyncApiKeyDelete,
+
+		Schema: map[string]*schema.Schema{
+			"description": {
+				Type:     schema.TypeString,
+				Optional: true,
+				Default:  "Managed by Terraform",
+			},
+			"api_id": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"expires": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				ValidateFunc: validation.ValidateRFC3339TimeString,
+			},
+			"key": {
+				Type:      schema.TypeString,
+				Computed:  true,
+				Sensitive: true,
+			},
+		},
+	}
+}
+
+func resourceAwsAppsyncApiKeyCreate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).appsyncconn
+	params := &appsync.CreateApiKeyInput{
+		ApiId:       aws.String(d.Get("api_id").(string)),
+		Description: aws.String(d.Get("description").(string)),
+	}
+	if v, ok := d.GetOk("expires"); ok {
+		t, _ := time.Parse(time.RFC3339, v.(string))
+		params.Expires = aws.Int64(t.Unix())
+	}
+	resp, err := conn.CreateApiKey(params)
+	if err != nil {
+		return err
+	}
+
+	d.SetId(fmt.Sprintf("%s:%s", d.Get("api_id").(string), *resp.ApiKey.Id))
+	return resourceAwsAppsyncApiKeyRead(d, meta)
+}
+
+func resourceAwsAppsyncApiKeyRead(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).appsyncconn
+	var listKeys func(*appsync.ListApiKeysInput) (*appsync.ApiKey, error)
+	ApiId, Id, er := decodeAppSyncApiKeyId(d.Id())
+	if er != nil {
+		return er
+	}
+	listKeys = func(input *appsync.ListApiKeysInput) (*appsync.ApiKey, error) {
```

Comment:
This should be made into its own top level function, so it can be used by `testAccCheckAwsAppsyncApiKeyDestroy` and `testAccCheckAwsAppsyncApiKeyExists` so they properly handle `NextToken`